### ISSUE TITLE
A new release and a CHANGELOG?

### ISSUE THREAD

**Issue by jeremywadsack** [abonas/kubeclient#281]
The releases on `rubygems` are confusing. 

```
2.5.1 - October 12, 2017 (47.5 KB)
2.4.0 - May 10, 2017 (47 KB)
2.3.0 - December 05, 2016 (46 KB)
```

Version 2.5.1 says it was released on 2017-10-12, but the [tag shows](https://github.com/abonas/kubeclient/tree/v2.5.1) the [latest commit](https://github.com/abonas/kubeclient/commit/c3783eac3a2a2357ea98c6fb90c71cbe437a42fb) was 2017-05-09.

Lacking a `CHANGELOG` or `HISTORY` document, it's hard to understand what changes are in each version and we're left with comparing releases and reading the commit statements (e.g. [2.3.0 to 2.5.1](https://github.com/abonas/kubeclient/compare/v2.3.0...v2.5.1)).

It would be really helpful to have a log of changes for each release. :)

It would  also be nice to have a new release as there have been [lots of changes](https://github.com/abonas/kubeclient/compare/v2.5.1...master), some of which were merged over a year ago.

**Comment by moolitayer** [comment]
@jeremywadsack our gem release process is undergoing change. 
We will take all feedback you can give under consideration and hopefully draft some rules for release related procedures.  

We plan a new release, hope I can update on that soon.

CC @cben

**Comment by cben** [comment]
One question I'd love feedback on is how useful a "stable" release branch with cherrypicks is at all?
A lot of time has passed since kubeclient had a release from master (and I agree it's high time to get all the changes out).  The default mode was "release when asked, with only the requested feature cherrypicked".
I wonder if default mode could switch to "release from master", with branching only to backport bug fixes, when requested.  I'd prefer to not backport features, unless someone has a specific need.

**Comment by cben** [comment]
> Version 2.5.1 says it was released on 2017-10-12, but the [tag shows](https://github.com/abonas/kubeclient/tree/v2.5.1) the latest commit was 2017-05-09.

Yeah, that last commit is confusing.
> commit c3783eac3a2a2357ea98c6fb90c71cbe437a42fb
> AuthorDate: Tue May 9 16:12:05 2017 +0200
> CommitDate: Thu Oct 12 11:06:17 2017 +0200
> 
>     Bump version to 2.5.1

Too bad GitHub UI doesn't show CommitDate when != AuthorDate.  If you look at history prior to that commit https://github.com/abonas/kubeclient/commits/v2.5.1 you'll see the merged commits are recent.

**Comment by cben** [comment]
Interim update:
We started going over all PRs that are on `master` and not yet released (== on `v2.x` branch).  There are quite a lot, and a few change public API.  So:

- We should release a 2.6 with all backward-compatible changes.
- We'll then look at releasing the remainder as 3.0 "soonish", but there are several more things that were discussed for 3.0 (#199) that might be worth sneaking in...

We started labeling PRs with "v2.x/yes" and "v2.x/no".  I'll post a summary when that's done.

**Comment by jeremywadsack** [comment]
@cben That sounds great. Thanks for all the work on this (to the whole team).

To your earlier question (that I somehow missed), I'm not sure I understand what you're asking, but I'm preferential to regular releases, with semantic versioning. If I don't want to address breaking changes on my end, I can always pin the dependency in my `.gemspec` or `Gemfile`. Backporting bugfixes is nice, although most important would be backporting security issues. Even then you'll have to decide on an EOL for a major version at some point.

**Comment by cben** [comment]
After some indecision on #289 on small vs large compatible release, we settled on releasing master as 3.0 first.
See https://github.com/cben/kubeclient/blob/changelog/CHANGELOG.md (#297) for draft changelog.

**Comment by cben** [comment]
If anyone has specific preferences on what you want in 2.x release(s), please share.

**Comment by cben** [comment]
tagged & released v3.0.0 :tada:

**Comment by cben** [comment]
Also working on small v2.5.2 release: #287, #301, #302.

The fixes we'll include in v2.5.2:
- Watch results are now `RecursiveOpenStruct` inside arrays too (#279).
- Fixed watch `.finish` sometimes caused `Errno::EBADF` exception from the reading loop (#280).
- Easing dependency version (#287, #301).

I'm shelving the large 2.6 attempt (#289) for now.
If anyone wants additional 2.x release, let us know what you need.

**Comment by cben** [comment]
tagged & released v2.5.2 :tada:
